File: flink_scala/src/main/scala/github/yahuili1128/streamOperators/IterateDemo.scala
package github.yahuili1128.streamOperators

import github.yahuili1128.connector.ScalaSourceKafka010
import github.yahuili1128.connector.ScalaSourceKafka010.MockUp
import org.apache.flink.streaming.api.scala._

/**
  * @Description : flink iterate scala
  * @Author : LiYahui
  * @Date : 2019-07-15 18:40
  * @Version : V1.0
  */
object IterateDemo {
	def main(args: Array[String]): Unit = {
		val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
		val mockup: DataStream[ScalaSourceKafka010.MockUp] = ScalaSourceKafka010.getMockUpFromKafka(env)
		val iterateMockupDS: DataStream[MockUp] = mockup.iterate {
			iteration => {
				val addAgeDS: DataStream[MockUp] = iteration.map(line => {
					MockUp(line.name, line.gender, line.timestamp, line.age + 2)
				})
				//		feedback , 过滤条件
				(addAgeDS.filter(_.age > 50), addAgeDS.filter(_.age < 40))
			}
		}

		iterateMockupDS.print().setParallelism(1)

		env.execute(s"${this.getClass.getSimpleName}")
	}
}

File: flink_java/src/main/java/github/yahuili1128/streamOperators/IterateDemo.java
package github.yahuili1128.streamOperators;

import github.yahuili1128.connector.SourceKafka010;
import github.yahuili1128.pojo.MockUpModel;
import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.IterativeStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * @Description : flink  iterate java
 * @Author : LiYahui
 * @Date : 2019-07-15 13:46
 * @Version : V1.0
 */
public class IterateDemo {
	public static void main(String[] args) throws Exception {
		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
		SingleOutputStreamOperator<MockUpModel> mockUpkafka010 = SourceKafka010.getMockUpkafka010(env);
		IterativeStream<MockUpModel> iterativeStream = mockUpkafka010.iterate();
		SingleOutputStreamOperator<MockUpModel> mockMapDS = iterativeStream
				.map(new MapFunction<MockUpModel, MockUpModel>() {
					@Override
					public MockUpModel map(MockUpModel value) throws Exception {
						MockUpModel mock = new MockUpModel();
						mock.setGender(value.gender);
						mock.setName(value.name);
						mock.setAge(value.age + 2);
						mock.setTimestamp(value.timestamp);
						return mock;
					}
				});
		SingleOutputStreamOperator<MockUpModel> feedback = mockMapDS.filter(new FilterFunction<MockUpModel>() {
			@Override
			public boolean filter(MockUpModel value) throws Exception {

				return value.age > 50;
			}
		});
		iterativeStream.closeWith(feedback);
		SingleOutputStreamOperator<MockUpModel> resultDS = mockMapDS.filter(new FilterFunction<MockUpModel>() {
			@Override
			public boolean filter(MockUpModel value) throws Exception {
				return value.age < 40;
			}
		});
		resultDS.print().setParallelism(1);

		env.execute("kafka010 demo");
	}
}
